Reliable Topic Source

ufukyilmaz opened the pull request:
PR closed by Hazelcast automation as no activity (>3 months). Please reopen with comments, if necessary. Thank you for using Hazelcast and your valuable contributions

ufukyilmaz committed:
Add ITopic Source Connector.

ufukyilmaz committed:
Rename the module. Add javadoc and test.

Commit:
Add reliable topic source

# Conflicts:
#	settings.gradle

Commit:
Add module info to README

viliam-durina commented on reliabletopic/src/main/java/com/hazelcast/jet/contrib/reliabletopic/ReliableTopicSource.java:
```diff
@@ -0,0 +1,86 @@
+            @Nonnull String topic,
+            @Nonnull FunctionEx<Message<M>, T> projectionFn
+    ) {
+        return SourceBuilder.timestampedStream("reliable-topic-source",
```
the name should include the topic name

viliam-durina commented on reliabletopic/src/main/java/com/hazelcast/jet/contrib/reliabletopic/ReliableTopicSource.java:
```diff
@@ -0,0 +1,86 @@
+        private final FunctionEx<Message<M>, T> projectionFn;
+
+        private ITopicSourceContext(Context ctx, String topic, @Nonnull FunctionEx<Message<M>, T> projectionFn) {
+            ctx.jetInstance().<M>getReliableTopic(topic).addMessageListener(queue::add);
```
1. `ArrayDeque` is non-concurrent, we must use a concurrent queue because the message listener adds to it in a different thread. A SPSC queue is probably sufficient, as per the specification of `MessageListener`.

2. We don't apply any upper bound on the queue. If the job is stuck or slow, the queue could grow big. Not sure how or if can we apply backpressure, needs to be investigated and/or documented.

ufukyilmaz replied:
1) I need to use thread-safe collection here 👍 . I also talked to Ali about this but we have trouble about which thread-safe collection I should choose here. Whether I need to choose bounded one or unbounded one? The reliable topic is already using RingBuffer, which is an internally bounded collection. If I want to use bounded collection, I need to either apply backpressure or use something like blocking queue(which is against Jet's policy on blocking calls). Also, I cannot apply backpressure on this message listener logic, [here](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java#L141-L144). Since we cannot apply backpressure, it is possible to encounter problems such as OOM if we use unbounded collections as you already said.
2) I stopped using the blocking queue just because the thread was blocked. However, although I remove to use a collection having a blocking operation, I am still faced with a blocking call in this source. The thread is blocked again when no message is sent to the topic, since the messageListener that is added to the topic read from this topic in a blocking manner, [here](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java#L89-L90). I think this is a problem for us.

-> This reliable topic is using internally ring buffer maybe we can implement our own reading logic from ring buffer.

viliam-durina commented on reliabletopic/src/main/java/com/hazelcast/jet/contrib/reliabletopic/ReliableTopicSource.java:
```diff
@@ -0,0 +1,86 @@
+
+    /**
+     * Creates a {@link StreamSource} which reads messages from
+     * from a Hazelcast reliable topic with the specified name.
```
```suggestion
     * from a Hazelcast Reliable Topic with the specified name.
```

viliam-durina commented on reliabletopic/src/main/java/com/hazelcast/jet/contrib/reliabletopic/ReliableTopicSource.java:
```diff
@@ -0,0 +1,86 @@
+     * @param <T> type of emitted item after projection.
+     * @return {@link StreamSource}
+     */
+    public static <M, T> StreamSource<T> topicSource(
```
`HazelcastInstance` has 2 methods:
- `ITopic getTopic()`
- `ITopic getReliableTopic()`

As far as I understand these don't share the namespace, but have the same interface. Because of this maybe we should support both (or we do already support the other?). The javadoc doesn't specify any difference between them... But we should be clear which one we mean and use the terminology. So this method should probably be named `reliableTopicSource`.

viliam-durina commented on reliabletopic/src/main/java/com/hazelcast/jet/contrib/reliabletopic/ReliableTopicSource.java:
```diff
@@ -0,0 +1,86 @@
+     */
+    public static <M, T> StreamSource<T> topicSource(
+            @Nonnull String topic,
+            @Nonnull FunctionEx<Message<M>, T> projectionFn
```
We should have an override without the `projectionFn`, it's quite common use case. Maybe also an override which will accept an `ITopic` - but beware, it can be both reliable and _non reliable_ topic.

ufukyilmaz replied:
I removed projectionFn completely.

viliam-durina commented on reliabletopic/src/main/java/com/hazelcast/jet/contrib/reliabletopic/ReliableTopicSource.java:
```diff
@@ -0,0 +1,86 @@
+import java.util.Queue;
+
+
+
```
Extra lines

viliam-durina commented on reliabletopic/src/main/java/com/hazelcast/jet/contrib/reliabletopic/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ */
+
+/**
+ * Connector for Hazelcast Jet to connect to IMDG topics.
```
```suggestion
 * Connector for Hazelcast Jet to connect to IMDG Reliable Topic.
```

ufukyilmaz committed:
Update reliabletopic/src/main/java/com/hazelcast/jet/contrib/reliabletopic/ReliableTopicSource.java

Co-authored-by: Viliam Durina <viliam-durina@users.noreply.github.com>

ufukyilmaz committed:
Update reliabletopic/src/main/java/com/hazelcast/jet/contrib/reliabletopic/package-info.java

Co-authored-by: Viliam Durina <viliam-durina@users.noreply.github.com>

Commit:
Use BlockingQueue to buffer items

Remove projectionFn to provide more convenient API
Change the name of API's
Update the tests

Commit:
Merge branch 'itopic-connector' of https://github.com/ufukyilmaz/hazelcast-jet-contrib into itopic-connector

 Conflicts:
	reliabletopic/src/main/java/com/hazelcast/jet/contrib/reliabletopic/ReliableTopicSource.java

Commit:
Small change on test

Commit:
Change source name to include topic name

devOpsHazelcast commented:
[![CLA assistant check](https://cla.hazelcast.org:443/pull/badge/not_signed)](https://cla.hazelcast.org:443/hazelcast/hazelcast-jet-contrib?pullRequest=110) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.hazelcast.org:443/hazelcast/hazelcast-jet-contrib?pullRequest=110) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: ufukyilmaz<br/>:x: Ufuk<br/><hr/>**Ufuk** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.hazelcast.org:443/check/hazelcast/hazelcast-jet-contrib?pullRequest=110) it.</sub>

devOpsHazelcast commented:
Can one of the admins verify this patch?
